File: tools/extract_i18n.js
const fs = require('fs');
const glob = require('fs').readdirSync;
const path = require('path');

const traverseDirs = (dir, ext) => {
    let results = [];
    glob(dir).forEach(file => {
        file = path.join(dir, file);
        const stat = fs.statSync(file);
        if (stat && stat.isDirectory()) { 
            if (!file.includes('vendor')) {
                results = results.concat(traverseDirs(file, ext));
            }
        } else if (file.endsWith(ext)) {
            results.push(file);
        }
    });
    return results;
};

const htmlFiles = traverseDirs('src/ui', '.html');
const jsFiles = traverseDirs('src/ui', '.js');

const keys = new Set();

const htmlRegex = /data-i18n(?:-[a-z]+)?="([^"]+)"/g;
htmlFiles.forEach(file => {
    const content = fs.readFileSync(file, 'utf8');
    let match;
    while ((match = htmlRegex.exec(content)) !== null) {
        keys.add(match[1]);
    }
});

const jsRegex = /(?<![\w])t\(['"]([^'"]+)['"]\)/g;
jsFiles.forEach(file => {
    const content = fs.readFileSync(file, 'utf8');
    let match;
    while ((match = jsRegex.exec(content)) !== null) {
        keys.add(match[1]);
    }
});

const sortedKeys = Array.from(keys).sort();

const constructObj = (keys) => {
    const obj = {};
    keys.forEach(k => {
        const parts = k.split('.');
        let current = obj;
        for (let i = 0; i < parts.length - 1; i++) {
            if (!current[parts[i]]) current[parts[i]] = {};
            current = current[parts[i]];
        }
        current[parts[parts.length - 1]] = k;
    });
    return obj;
};

const res = constructObj(sortedKeys);
fs.writeFileSync('src/ui/i18n/en.json', JSON.stringify(res, null, 2));
fs.writeFileSync('src/ui/i18n/es.json', JSON.stringify(res, null, 2));
console.log('Created en.json and es.json');

File: tools/extract_i18n.py
import os, re, json

def find_files(d, ext):
    res = []
    for root, dirs, files in os.walk(d):
        if 'vendor' in root: continue
        for f in files:
            if f.endswith(ext): res.append(os.path.join(root, f))
    return res

keys = set()

for f in find_files('src/ui', '.html'):
    with open(f, 'r', encoding='utf-8') as file:
        content = file.read()
        for m in re.finditer(r'data-i18n(?:-[a-z]+)?="([^"]+)"', content):
            keys.add(m.group(1))

for f in find_files('src/ui', '.js'):
    with open(f, 'r', encoding='utf-8') as file:
        content = file.read()
        for m in re.finditer(r'(?<![\w])t\([\'"]([^\'"]+)[\'"]\)', content):
            keys.add(m.group(1))

obj = {}
for k in sorted(list(keys)):
    parts = k.split('.')
    current = obj
    for p in parts[:-1]:
        if p not in current: current[p] = {}
        current = current[p]
    current[parts[-1]] = k

os.makedirs('src/ui/i18n', exist_ok=True)
with open('src/ui/i18n/en.json', 'w', encoding='utf-8') as file:
    json.dump(obj, file, indent=2)

with open('src/ui/i18n/es.json', 'w', encoding='utf-8') as file:
    json.dump(obj, file, indent=2)

print('Done')
